File: Quicksort.java
// December 2, 2023

import java.util.Random;
public class Quicksort {

   public static int partition(int[] arr, int lowIndex, int highIndex) {
      int midPoint = (lowIndex + highIndex) / 2;
      int pivot = arr[midPoint];
      // System.out.println("pivot: " + pivot);

      boolean done = false;

      while (!done) {
         while (arr[lowIndex] < pivot)
            lowIndex++;
         while (arr[highIndex] > pivot)
            highIndex--;
         if (lowIndex >= highIndex) {
            done = true;
         } else {
            int temp = arr[lowIndex];
            arr[lowIndex] = arr[highIndex];
            arr[highIndex] = temp;

            lowIndex++;
            highIndex--;
         }
      }
      return highIndex;
   }
   
   public static void quicksortHelper(int[] arr, int lowIndex, int highIndex) {
      if (lowIndex >= highIndex) {
         return;
      }
      int lowEndIndex = partition(arr, lowIndex, highIndex);
      quicksortHelper(arr, lowIndex, lowEndIndex);
      quicksortHelper(arr, lowEndIndex + 1, highIndex);

   }

   // Method to generate a random array
   public static int[] generateRandomArray(int size) {
      int[] array = new int[size];
      Random random = new Random();

      for (int i = 0; i < size; i++) {
         array[i] = random.nextInt(1000000); // Change the range of random numbers as per your requirement
      }

      return array;
   }

   public static void main(String args[]){
      int[] arr = generateRandomArray(10);
      System.out.println("Unsorted array: ");
      for (int i = 0; i < arr.length; i++) {
         System.out.print(arr[i]+ " ");
      }
      quicksortHelper(arr, 0, arr.length - 1);
      System.out.println("\nSorted array: ");
      for (int i = 0; i < arr.length; i++) {
         System.out.print(arr[i]+ " ");
      }
      System.out.println();
   }

}

File: Quicksort.py
import random

def main():
   size = int(input("what size: "))
   a = []

   for i in range(0, size):
      a.append(int(random.random()*1000+1))

   print("Original Array: "+str(a))

   quicksort(a, 0, len(a)-1)

   print("Sorted Array: "+str(a))


def quicksort(a, low, high):
   if low < high:
      end = partition(a, low, high)
      quicksort(a, low, end)
      quicksort(a, end + 1, high)

def partition(a, low, high):
   mid = int((low+high)/2)
   pivot = a[mid]

   done = False
   while(not(done)):
      while(a[low] < pivot):
         low+=1
      while(a[high] > pivot):
         high-=1
      
      if (low >= high):
         done = True
      else:
         a[low], a[high] = a[high], a[low]
         low+=1
         high-=1
   
   return high

main()